packages/react-email/src/cli/utils/preview/hot-reloading/create-dependency-graph.ts

```diff
@@ -0,0 +1,244 @@
+ */
+export const createDependencyGraph = async (directory: string) => {
+  const filePaths = await readAllFilesInsideDirectory(directory);
+  const modulePaths = filePaths.filter(isJavascriptModule);
+  const graph: DependencyGraph = Object.fromEntries(
+    modulePaths.map((path) => [
+      path,
+      {
+        path,
+        dependencyPaths: [],
+        dependentPaths: [],
+        moduleDependencies: [],
+      },
+    ]),
+  );
+
+  const getDependencyPaths = async (filePath: string) => {
+    const contents = await fs.readFile(filePath, 'utf8');
+
+    const importedPaths = getImportedModules(contents);
+    const importedPathsRelativeToDirectory = importedPaths.map(
+      (dependencyPath) => {
+        const isModulePath = !dependencyPath.startsWith('.');
+        /*
+          path.isAbsolute will return false if the path looks like JavaScript module imports
+          e.g. path.isAbsolute('react-dom/server') will return false, but for our purposes this
+          path is not a relative one.
+        */
+        if (!isModulePath && !path.isAbsolute(dependencyPath)) {
+          let pathToDependencyFromDirectory = path.resolve(
+            /*
+              path.resolve resolves paths differently from what imports on javascript do.
+
+              So if we wouldn't do this, for an email at "/path/to/email.tsx" with a dependecy path of "./other-email" 
+              would end up going into /path/to/email.tsx/other-email instead of /path/to/other-email which is the
+              one the import is meant to go to
+            */
+            path.dirname(filePath),
+            dependencyPath,
+          );
+
+          /*
+            If the path to the dependency does not include a file extension, such that our check
+            for it being a javascript module fails, then we can assume it has the same as the `filePath`
+          */
+          if (!isJavascriptModule(pathToDependencyFromDirectory)) {
+            pathToDependencyFromDirectory = `${pathToDependencyFromDirectory}${path.extname(
+              filePath,
+            )}`;
+          }
+
+          return pathToDependencyFromDirectory;
+        } else {
+          // when either the path is a module or is absolute
+          return dependencyPath;
+        }
+      },
+    );
+
+    const moduleDependencies = importedPathsRelativeToDirectory.filter(
+      (dependencyPath) =>
+        !dependencyPath.startsWith('.') && !path.isAbsolute(dependencyPath),
+    );
+
+    const nonNodeModuleImportPathsRelativeToDirectory =
+      importedPathsRelativeToDirectory.filter(
+        (dependencyPath) =>
+          dependencyPath.startsWith('.') || path.isAbsolute(dependencyPath),
+      );
+
+    return {
+      dependencyPaths: nonNodeModuleImportPathsRelativeToDirectory,
+      moduleDependencies,
+    };
+  };
+
+  const updateModuleDependenciesInGraph = async (moduleFilePath: string) => {
```

Comment:
This creates and adds the node into the graph for the path if it doesn't already exist. If it does exist, then it updates the dependencies and then updates the references to it on the `dependentPaths` of other nodes.